### ISSUE TITLE
Add AdamW Optimiser

### ISSUE THREAD

**Issue by BradyPlanden** [pybop-team/PyBOP#295]
### Feature description

Add an AdamW implementation to PyBOP. To achieve this, an implementation that matches the ask-tell interface should be completed. This optimiser will likely be the first pure PyBOP optimisation method.

### Motivation

[AdamW](https://arxiv.org/abs/1711.05101) provides a weight decay implementation of the Adam algorithm. A pytorch implementation can be found here: https://pytorch.org/docs/stable/generated/torch.optim.AdamW.html
